[CALCITE-6283] Function array_append with a NULL array argument crashes with NullPointerException

mihaibudiu opened the pull request:
…

tanclary commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -6268,6 +6268,18 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
+    f.checkNull("array_append(null, 2)");
+    f.checkNull("array_prepend(null, 2)");
+    f.checkNull("array_remove(null, 2)");
+    f.checkNull("array_contains(null, 2)");
```
Do we have tests for appending null elements

caicancai replied:
@tanclary Yes, similar to:
 f.checkNull("array_append(cast(null as integer array), 1)");

tanclary replied:
I meant a test where the element is null, not the array. But i found that we do so that's good.

caicancai replied:
I tested it in spark3.4. It seems that the first parameter of array_append is null. Why do we need to change it like this?

caicancai replied:
![捕获](https://github.com/apache/calcite/assets/77189278/a081d503-71a1-4222-b3f1-70a35b0e4a22)

tanclary replied:
I don't understand your question

caicancai replied:
My problem is that array_append is to adapt to spark, but spark’s array_append(null, 2) will report an error.

caicancai replied:
What I mean is that spark3.4 version does not support array_append(null, 1), it will report an error. Currently, array_append and array_prepend are adapted for spark in calcite. If spark does not support array_append(null , 1), why do we need to return null? I am not sure whether higher versions of spark support it.
I don't know if my opinion is useful, if it is problematic, please ignore it. But I still want to express my opinion. This is open source, thank you.
![捕获](https://github.com/apache/calcite/assets/77189278/dfefad4e-45d0-431a-a55c-df340878ba16)

tanclary replied:
Ah. Yes I agree with you. What do you think @mihaibudiu ?

mihaibudiu replied:
I have changed the type checker to reject NULL literals for arrays.
However, it seems to be that Spark is inconsistent, since it allows NULL values for arrays at runtime, but not at compile time. This is probably because Spark uses a different type checking/inference algorithm.

caicancai replied:
> I have changed the type checker to reject NULL literals for arrays. However, it seems to be that Spark is inconsistent, since it allows NULL values for arrays at runtime, but not at compile time. This is probably because Spark uses a different type checking/inference algorithm.

spark sql, after generating the Unresolved Logical Plan, runs a series of rules to turn it into the Logical Plan, which solves some type conversion issues there, and the type handling of array_append(null,2) seems to be in that part

tanclary commented on core/src/main/java/org/apache/calcite/sql/type/ArrayElementOperandTypeChecker.java:
```diff
@@ -45,23 +45,27 @@ public class ArrayElementOperandTypeChecker implements SqlOperandTypeChecker {
 
-    RelDataType arrayComponentType =
-        getComponentTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0));
+    RelDataType op0Type = SqlTypeUtil.deriveType(callBinding, op0);
```
can this be final

caicancai commented:
@mihaibudiu I have also recently been fixing the bug of array_append, which seems to conflict with this PR.
https://github.com/apache/calcite/pull/3705

chucheng92 commented:
pls also see https://issues.apache.org/jira/browse/CALCITE-5893 to support array_except/array_intersect.  they also should reject NULL literals.

chucheng92 commented:
FYI: One background is that many spark functions related to calcite's array are contributed by one contributor. One basic strategy is to turn on NullPolicy, which means that NULL literals, including **NULL** and **cast(null as xxx)** are returned null. But in fact, spark will throw an exception in the former. We need to address the former case.

mihaibudiu commented:
The situation may be even more complicated if one considers cases where the compiler needs to insert implicit casts. I don't have a spark engine handy to test what Spark does, but if the array element types and the added element do not have the same type, Spark may choose to insert implicit casts in different ways than Calcite does. 

Another fact I don't have a good handle on is the nullability of array elements: https://issues.apache.org/jira/browse/CALCITE-6275

chucheng92 commented:
array_append needs to be capitalized in commit name/jira name/pr name.

chucheng92 commented:
> The situation may be even more complicated if one considers cases where the compiler needs to insert implicit casts. I don't have a spark engine handy to test what Spark does, but if the array element types and the added element do not have the same type, Spark may choose to insert implicit casts in different ways than Calcite does.
> 
> Another fact I don't have a good handle on is the nullability of array elements: https://issues.apache.org/jira/browse/CALCITE-6275

make sense. I just read it wrong. Your PR only handles the situation when array is null.

mihaibudiu commented:
> make sense. I just read it wrong. Your PR only handles the situation when array arg is NULL. My point is that the element is also a NULL literal.

There are already tests for this case, e.g.: `testArrayAppendFunc`

mihaibudiu commented:
> pls also see https://issues.apache.org/jira/browse/CALCITE-5893 to support array_except/array_intersect. they also should reject NULL literals.
> 
> calcite spark:
> 
> // return null select array_contains(array[1, 2], null); // return null select array_except(array[1, 2, 3], null) // return null select array_intersect(array[1,2,3], null)
> 
> they should reject NULL.

That's great, we should merge that PR if it solves the problem. I will try to review it.

mihaibudiu commented:
> array_append needs to be capitalized in commit name/jira name/pr name.

I will update these when I squash the commits

chucheng92 commented:
> > make sense. I just read it wrong. Your PR only handles the situation when array arg is NULL. My point is that the element is also a NULL literal.
> 
> There are already tests for this case, e.g.: `testArrayAppendFunc`

If the first array arg is NULL, then it is an exception for all these array functions. There is no doubt about this, but if the second arg is NULL, the semantics are different. `array_contains/array_except/array_intersect` is currently wrong. And you are right to mention `array_append/array_prepend` as it does insert null.

![image](https://github.com/apache/calcite/assets/11287509/1fc3f24a-a656-4259-bf3d-57878ecdda6f)

yes, I have fixed `array_contains/array_except/array_intersect` in that PR, and that PR will affect this PR.

mihaibudiu commented:
Let's merge [CALCITE-5893] first and I will rebase this one.

snuyanzin commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -6277,6 +6277,19 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
+  @Test void testArrayNullFunc() {
+    final String expected = "Illegal use of 'NULL'";
+    final SqlOperatorFixture f = fixture().withLibrary(SqlLibrary.SPARK);
+    f.checkFails("^array_append(null, 2)^", expected, false);
+    f.checkFails("^array_prepend(null, 2)^", expected, false);
+    f.checkFails("^array_remove(null, 2)^", expected, false);
+    f.checkFails("^array_contains(null, 2)^", expected, false);
+    f.checkFails("^array_position(null, 2)^", expected, false);
```
It seems there is a similar issue for `array_min/array_max` should we fix it within this issue or create a separate one?

mihaibudiu replied:
since I got started I will add these too

mihaibudiu commented:
In my last commit I have fixed array_min and array_max as well.
I have slightly modified the helper function introduced by @chucheng92 in https://github.com/apache/calcite/pull/3355 and used it for typechecking more functions.
I have also corrected a test for array_contains, according to the Databricks manual, that test was incorrect.
https://docs.databricks.com/en/sql/language-manual/functions/array_contains.html

chucheng92 commented on core/src/main/java/org/apache/calcite/sql/type/ArrayElementOperandTypeChecker.java:
```diff
@@ -34,61 +33,64 @@
 
   @Override public boolean checkOperandTypes(
       SqlCallBinding callBinding,
       boolean throwOnFailure) {
```
If array is not first argument? such as array_xxx(1, array, 2). It seems that this typechecker only works in this situation.

chucheng92 commented:
overall looks good to me +1

mihaibudiu commented:
Interpreting the +1 as a "go ahead and merge"
